remotes/promises.go

```diff
@@ -0,0 +1,189 @@
+package remotes
+
+import (
+	"context"
+	"reflect"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type scheduler interface {
+	schedule(func(ctx context.Context) error) promise
+	ctx() context.Context
+}
+
+type completionSource struct {
+	done chan struct{}
+	err  error
+}
+
+func (cs *completionSource) Done() <-chan struct{} {
+	return cs.done
+}
+
+func (cs *completionSource) Err() error {
+	return cs.err
+}
+
+func newPromise(scheduler scheduler, dependency dependency) promise {
+	return promise{scheduler: scheduler, dependency: dependency}
+}
+
+type promise struct {
+	dependency
+	scheduler scheduler
+}
+
+func (p promise) wait() error {
+	<-p.Done()
+	return p.Err()
+}
+
+func (p promise) then(next func(ctx context.Context) error) promise {
+	completionSource := &completionSource{
+		done: make(chan struct{}),
+	}
+	go func() {
+		defer close(completionSource.done)
+		select {
+		case <-p.Done():
+			if err := p.Err(); err != nil {
+				completionSource.err = err
+				return
+			}
+			completionSource.err = p.scheduler.schedule(next).wait()
+		}
+	}()
+	return newPromise(p.scheduler, completionSource)
+}
+
+type dependency interface {
+	Done() <-chan struct{}
+	Err() error
+}
+
+type failedDependency struct {
+	err error
+}
+
+func (f failedDependency) Done() <-chan struct{} {
+	return doneCh
+}
+
+func (f failedDependency) Err() error {
+	return f.err
+}
+
+type doneDependency struct {
+}
+
+func (doneDependency) Done() <-chan struct{} {
+	return doneCh
+}
+
+func (doneDependency) Err() error {
+	return nil
+}
+
+func whenAll(dependencies []dependency) dependency {
+	completionSource := &completionSource{
+		done: make(chan struct{}),
+	}
+	go func() {
+		defer close(completionSource.done)
+		cases := make([]reflect.SelectCase, len(dependencies))
+		for ix, dependency := range dependencies {
+			cases[ix] = reflect.SelectCase{
+				Chan: reflect.ValueOf(dependency.Done()),
+				Dir:  reflect.SelectRecv,
+			}
+		}
+		for len(dependencies) > 0 {
+			ix, _, _ := reflect.Select(cases)
+			if err := dependencies[ix].Err(); err != nil {
+				completionSource.err = err
+				return
+			}
+			cases = append(cases[:ix], cases[ix+1:]...)
+			dependencies = append(dependencies[:ix], dependencies[ix+1:]...)
+		}
+	}()
+	return completionSource
+}
+
+type todoItem struct {
+	completionSource *completionSource
+	do               func(ctx context.Context) error
+}
+
+type errgroupScheduler struct {
+	workGroup *errgroup.Group
+	todoList  chan todoItem
+	context   context.Context
+}
+
+// creates a scheduler that cancels all tasks at the first error
+func newErrgroupScheduler(ctx context.Context, workerCount, todoBuffer int) *errgroupScheduler {
+	todoList := make(chan todoItem, todoBuffer)
+	workGroup, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < workerCount; i++ {
+		workGroup.Go(func() error {
+			for {
+				select {
+				case todoItem := <-todoList:
+					todoItem.completionSource.err = todoItem.do(ctx)
+					close(todoItem.completionSource.done)
```

Comment:
`close()` called before setting the error.

Reply:
Actually it is not, it is just the early errgroup return that does that.